File: src/bin/0004-median-of-two-sorted-arrays.cpp
// @leetup=info id=4 lang=cpp slug=median-of-two-sorted-arrays

#include <algorithm>
#include <cassert>
#include <climits>
#include <cstdint>
#include <iostream>
#include <map>
#include <set>
#include <tuple>
#include <vector>

// @leetup=code

class Solution {
 public:
  double findMedianSortedArrays(std::vector<int>& nums1,
                                std::vector<int>& nums2) {
    if (nums1.size() > nums2.size()) {
      return findMedianSortedArrays(nums2, nums1);
    }

    size_t total_size = nums1.size() + nums2.size();
    size_t half_size = total_size / 2;

    size_t left = 0;
    size_t right = nums1.size();

    // This should be left <= right, instead of left < right because
    //  we need to run this loop for left==right at last.
    while (left <= right) {
      size_t mid1 = left + (right - left) / 2;
      size_t mid2 = half_size - mid1;

      int left1 = (mid1 >= 1) ? nums1[mid1 - 1] : INT_MIN;
      int right1 = (mid1 < nums1.size()) ? nums1[mid1] : INT_MAX;

      int left2 = (mid2 >= 1) ? nums2[mid2 - 1] : INT_MIN;
      int right2 = (mid2 < nums2.size()) ? nums2[mid2] : INT_MAX;

      if (left1 <= right2 && left2 <= right1) {
        if (total_size % 2 == 1) {
          return std::min(right1, right2);
        }
        return (std::max(left1, left2) + std::min(right1, right2)) / 2.0;
      }
      if (left1 < right2) {
        left = mid1 + 1;
      } else {
        right = mid1;
      }
    }
    assert(false);
  }
};
// @leetup=code

int main() {
  std::cout << "Hello, world!" << std::endl;
  return 0;
}

File: src/bin/0004-median-of-two-sorted-arrays.py
# @leetup=info id=4 lang=python3 slug=median-of-two-sorted-arrays



from typing import Optional

# @leetup=code

import math


class Solution:

    inf = math.inf

    def findMedianSortedArrays(self, nums1: list[int], nums2: list[int]) -> float:
        a = nums1
        b = nums2

        if len(a) > len(b):
            # Let len(a) < len(b) since b_mid (= half - a_mid) should be >= 0
            a, b = b, a

        total = len(a) + len(b)
        half = total // 2

        left = 0
        right = len(a)
        # This should be left <= right, instead of left < right because
        #  we need to run this loop for left==right at last.
        while left <= right:
            a_mid = (left + right) // 2

            # [x, x, x, ..]
            #        a_mid

            # A_left = A[:a_mid]
            # A_right = A[a_mid:]

            # [x, x, x, ..]
            #        b_mid

            # B_left = A[:b_mid]
            # B_right = A[b_mid:]

            # half = a_mid + b_mid

            b_mid = half - a_mid
            # print(f"{a_mid=}, {a[:a_mid]=}, {b_mid=}, {b[:b_mid]=}")

            a_left = a[a_mid - 1] if a_mid - 1 >= 0 else -inf
            a_right = a[a_mid] if a_mid < len(a) else inf

            b_left = b[b_mid - 1] if b_mid - 1 >= 0 else -inf
            b_right = b[b_mid] if b_mid < len(b) else inf

            if a_left <= b_right and b_left <= a_right:
                # print(f"{a[:a_mid]=}, {b[:b_mid]=}, {total=}, {half=}")
                # odd
                if total % 2:
                    return min(a_right, b_right)
                # even
                return (max(a_left, b_left) + min(a_right, b_right)) / 2
            elif a_left > b_right:
                right = a_mid
            else:
                left = a_mid + 1

        # Unreachable here


# @leetup=code
